plugins/algoltek-usbcr/fu-algoltek-usbcr-device.c

```diff
@@ -0,0 +1,671 @@
+		    {0xC7, 0x00},
+		    {0xC8, 0x07},
+		};
+		for (guint j = 0; j < G_N_ELEMENTS(regs); j++) {
+			if (!fu_algoltek_usbcr_device_write_reg(self,
+								regs[j].reg,
+								regs[j].val,
+								FU_AG_USBCR_RD_WR_XDATA,
+								error))
+				return FALSE;
+		}
+
+		back_data[0] = 0;
+		do {
+			if (!fu_algoltek_usbcr_device_read_reg(self,
+							       0xC8,
+							       back_data,
+							       1,
+							       FU_AG_USBCR_RD_WR_XDATA,
+							       error))
+				return FALSE;
+		} while (back_data[0] & 0x01);
+
+		back_data[0] = 0;
+		if (!fu_algoltek_usbcr_device_read_reg(self,
+						       0x400,
+						       back_data,
+						       2,
+						       FU_AG_USBCR_RD_WR_XDATA,
+						       error))
+			return FALSE;
+	} while (back_data[0] & 0x01);
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_algoltek_usbcr_device_write_chunks(FuAlgoltekUsbcrDevice *self,
+				      FuChunkArray *chunks,
+				      FuProgress *progress,
+				      GError **error)
+{
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_set_steps(progress, fu_chunk_array_length(chunks));
+
+	for (gint i = fu_chunk_array_length(chunks) - 1; i >= 0; i--) {
+		g_autoptr(FuChunk) chk = NULL;
+
+		chk = fu_chunk_array_index(chunks, i, error);
+		if (chk == NULL)
+			return FALSE;
+		if (!fu_algoltek_usbcr_device_write_chunk(self, chk, error))
+			return FALSE;
+		fu_progress_step_done(progress);
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_algoltek_usbcr_device_verify_chunks(FuAlgoltekUsbcrDevice *self,
+				       FuChunkArray *chunks,
+				       FuProgress *progress,
+				       GError **error)
+{
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_set_steps(progress, fu_chunk_array_length(chunks));
+
+	for (guint i = 0; i < fu_chunk_array_length(chunks); i++) {
+		g_autofree guint8 *buf = NULL;
+		g_autoptr(FuChunk) chk = NULL;
+
+		chk = fu_chunk_array_index(chunks, i, error);
+		if (chk == NULL)
+			return FALSE;
+		buf = g_malloc0(fu_chunk_get_data_sz(chk));
+		if (!fu_algoltek_usbcr_device_do_read_spi(self,
+							  fu_chunk_get_address(chk),
+							  buf,
+							  fu_chunk_get_data_sz(chk),
+							  error))
+			return FALSE;
+		if (!fu_memcmp_safe(fu_chunk_get_data(chk),
+				    fu_chunk_get_data_sz(chk),
+				    0,
+				    buf,
+				    fu_chunk_get_data_sz(chk),
+				    0x0,
+				    fu_chunk_get_data_sz(chk),
+				    error))
+			return FALSE;
+
+		fu_progress_step_done(progress);
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_algoltek_usbcr_device_write_firmware(FuDevice *device,
+					FuFirmware *firmware,
+					FuProgress *progress,
+					FwupdInstallFlags flags,
+					GError **error)
+{
+	FuAlgoltekUsbcrDevice *self = FU_ALGOLTEK_USBCR_DEVICE(device);
+	g_autoptr(GInputStream) stream = NULL;
+	g_autoptr(FuChunkArray) chunks = NULL;
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_flag(progress, FU_PROGRESS_FLAG_GUESSED);
```

Comment:
@lingc24 do you get output from the daemon in reed that say things like "steps were set as [ 10 80 10 ] but should have been [ 1 97 2 ]"? if you fill in these numbers and remove `FU_PROGRESS_FLAG_GUESSED` then the daemon gets less mad at you.

Reply:
Updated, remove this line of FU_PROGRESS_FLAG_GUESSED.